src/replication.c

```diff
@@ -500,6 +499,59 @@ trigger_rebuild(spec_t *spec)
 	}
 }
 
+static bool
+is_replica_familiar(spec_t *spec, replica_t *new_replica)
+{
+	known_replica_t *kr = NULL;
+	int familiar_replicas = 0;
+	bool ret = false, found = false;
+	replica_t *old_replica = NULL;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	TAILQ_FOREACH(kr, &spec->identified_replica, next) {
+		familiar_replicas++;
+		if (kr->zvol_guid == new_replica->zvol_guid) {
+			found = true;
+			if (!kr->is_connected) {
+				ret = true;
+				kr->is_connected = true;
+			}
+			break;
+		}
+	}
+
+	if (!found && familiar_replicas < spec->replication_factor) {
+		kr = malloc(sizeof (known_replica_t));
+		kr->zvol_guid = new_replica->zvol_guid;
+		kr->is_connected = true;
+		ret = true;
+		TAILQ_INSERT_TAIL(&spec->identified_replica, kr, next);
+	}
+
+	/*
+	 * If the target has already an active session with this replica
+	 * then the target will disconnect from the previous session and
+	 * allow a new session
+	 */
+	if (found && !ret) {
```

Comment:
let us continue the old one, and error out the new connection due to complexities in looking for any pending IOs / pending epoll signals.
Is there any reason why we should consider new connection?

Reply:
let us continue the old one, and error out the new connection due to complexities in looking for any pending IOs / pending EPOLL signals.
>Good find! We can avoid new connection if we set keepalive for TCP socket. I have created an issue(**US2827**) for the same.

Is there any reason why we should consider a new connection?
>If the system(replica node) crash happens then target won't get any event regarding replica disconnect. Once system/replica comes up then it will try to make a new connection. Now, the target will have two sessions from the same replica.
As of now, the old session will take some time for disconnect event since we are not setting keepalive for a socket. Due to this, we are considering the new connection.